server/src/main/java/org/opensearch/index/store/RemoteSegmentStoreDirectory.java

```diff
@@ -769,48 +836,60 @@ public void deleteStaleSegments(int lastNMetadataFilesToKeep) throws IOException
+            sortedMetadataFileList,
+            allLockFiles,
+            lastNMetadataFilesToKeep
+        );
+        Map<String, String> activeNextMetadata = populateNextMetadata(sortedMetadataFileList, allLockFiles, lastNMetadataFilesToKeep);
         Set<String> deletedSegmentFiles = new HashSet<>();
+        List<String> metadataFilesWithSuccessfulSegmentDeletion = new ArrayList<>();
         for (String metadataFile : metadataFilesToBeDeleted) {
             Map<String, UploadedSegmentMetadata> staleSegmentFilesMetadataMap = readMetadataFile(metadataFile).getMetadata();
             Set<String> staleSegmentRemoteFilenames = staleSegmentFilesMetadataMap.values()
                 .stream()
                 .map(metadata -> metadata.uploadedFilename)
                 .collect(Collectors.toSet());
+
+            processActiveSegments(
+                activePreviousMetadata.get(metadataFile),
+                activeSegmentFilesMetadataMap,
+                activeSegmentRemoteFilenames,
+                metadataAlreadyProcessed
+            );
+            processActiveSegments(
+                activeNextMetadata.get(metadataFile),
+                activeSegmentFilesMetadataMap,
+                activeSegmentRemoteFilenames,
+                metadataAlreadyProcessed
+            );
             AtomicBoolean deletionSuccessful = new AtomicBoolean(true);
+
+            List<String> staleSegmentFilesToBeDeleted = new ArrayList<>();
             staleSegmentRemoteFilenames.stream()
                 .filter(file -> activeSegmentRemoteFilenames.contains(file) == false)
                 .filter(file -> deletedSegmentFiles.contains(file) == false)
                 .forEach(file -> {
-                    try {
-                        remoteDataDirectory.deleteFile(file);
-                        deletedSegmentFiles.add(file);
-                        if (!activeSegmentFilesMetadataMap.containsKey(getLocalSegmentFilename(file))) {
-                            segmentsUploadedToRemoteStore.remove(getLocalSegmentFilename(file));
-                        }
-                    } catch (NoSuchFileException e) {
-                        logger.info("Segment file {} corresponding to metadata file {} does not exist in remote", file, metadataFile);
-                    } catch (IOException e) {
-                        deletionSuccessful.set(false);
-                        logger.warn(
-                            "Exception while deleting segment file {} corresponding to metadata file {}. Deletion will be re-tried",
-                            file,
-                            metadataFile
-                        );
+                    staleSegmentFilesToBeDeleted.add(file);
+                    deletedSegmentFiles.add(file);
+                    if (!activeSegmentFilesMetadataMap.containsKey(getLocalSegmentFilename(file))) {
+                        segmentsUploadedToRemoteStore.remove(getLocalSegmentFilename(file));
```

Comment:
was checking more on this.. so in cases of primary failover also, we will initialize this cache with the segment files of latest metadata present in remote store. so since we anyway do not guarantee that this cache will be totally in sync with segments present in remote store. even with batch deletes we are not changing any behavior as such.

Reply:
@sohami let me know what are your thoughts on this. If we all agree i will pull the batch deletion changes in a follow up PR.

Reply:
@gbbafna @harishbhakuni I am not sure if the issue in existing logic w.r.t cache consistency is something known or not. My call out is with batch delete mechanism we are increasing the cases for cache being inconsistent. I agree that batch delete will help with the performance of delete calls, but we should probably implement it keeping the cache in mind. I don't fully know how this cache is being used and what are the guarantees the user of this cache expect from it, so we should evaluate that. If there is no issue for making this cache out of sync with batch logic, then it is fine to make that change in separate PR but the question will be then why this cache is being even updated in the delete path and why it is needed at all.

Reply:
thanks @sohami will create another issue to follow up on this.

Reply:
> @gbbafna @harishbhakuni I am not sure if the issue in existing logic w.r.t cache consistency is something known or not. My call out is with batch delete mechanism we are increasing the cases for cache being inconsistent. I agree that batch delete will help with the performance of delete calls, but we should probably implement it keeping the cache in mind. I don't fully know how this cache is being used and what are the guarantees the user of this cache expect from it, so we should evaluate that. If there is no issue for making this cache out of sync with batch logic, then it is fine to make that change in separate PR but the question will be then why this cache is being even updated in the delete path and why it is needed at all.

It is used mainly for segments which are referred to by the current segment_N file. For stale segments, we won't refer the cache for it . It is okay to contain extra files which are already deleted & not referenced , but not okay to miss a file which is already uploaded . 

1. fileLength  - Used to see the progress of recovery etc . Used for latest segments only 
2. is remote store in sync - Instead of list call we use this to find whatever is uploaded to remote . Used for latest segments only  
3. isRefreshAfterCommit - Figoure out if there is new segment N

>  but the question will be then why this cache is being even updated in the delete path and why it is needed at all. 

If we don't delete it, the size of the map can grow indefinitely. This would also be usefule in having another version of `deleteStaleSegments` which does a full clean up again on failed-to-delete segments .